### PR TITLE
Avoid clashing class_ slot in List parameter

### PR DESCRIPTION
**Pull request by jlstevens**
Here is one way to address https://github.com/holoviz/param/issues/455, namely to make sure the `List` parameter does not use the same `class_` slot as `ClassSelector` (I called it `list_class` for now instead).

Generally, I'm not sure why parameters inherit previous slot values at all (when they have a value of None) tbh...

### CODE REVIEW

**Comment by jlstevens** [comment]
This is probably the real issue: https://github.com/holoviz/param/issues/97

**Comment by jlstevens** [comment]
The tests are passing and this is a fairly simple workaround for #97. Even then, I would argue there is a difference between `class_` in class selector and `value_class` for the values in collections (lists, sets, dicts etc) justifying a separate slot.

**Comment by philippjfr** [comment]
I'd agree it justifies a different slot but as you say this is really a workaround for a deeper issue. Also not sure about `list_class` as the name, I'd suggest `item_type` but don't like that much more. In the long run I'd like towards making Parameters composable rather than having `item_type`-like slots.

**Comment by jlstevens** [comment]
Happy to rename to `item_type` and I also agree that a compositional model would be better for types. After the renaming would you be happy with this PR due to the difference in semantics or would you say #97 needs to be addressed first (which is bound to be a much bigger job)?

**Comment by philippjfr** [comment]
Maybe @jbednar has a better suggestion for the naming. I'd at least like to see `class_` to be aliased properly so whatever new name we decide on is also accepted as a kwarg.

**Comment by jbednar** [comment]
#97 does need addressing, but I agree it's a much larger job than putting in a workaround here, so we should do both. Regarding this specific workaround, I agree that the type for the items in a collection is semantically different than the type for a selector, so having a separate slot is very reasonable.  The name `item_type` conveys to me that it's the type of the item and not of the collection, so I'm happy with that.

File: param/__init__.py [hunk 1]
```diff
@@ -1359,11 +1359,11 @@ class List(Parameter):
     items in the list are checked to be of that type.
     """
 
-    __slots__ = ['class_','bounds']
+    __slots__ = ['bounds', 'item_type']
 
-    def __init__(self,default=[],class_=None,instantiate=True,
+    def __init__(self,default=[],class_=None,item_type=None,instantiate=True,
                  bounds=(0,None),**params):
-        self.class_ = class_
```
Comment (jlstevens) [inline param/__init__.py:1366]: @jbednar @philippjfr Think it is worth adding a warning to tell people to use `item_type` here instead if they set `class_` or would that be too annoying/noisy?

Comment (jbednar) [reply param/__init__.py:1366]: Too annoying.

**Comment by jlstevens** [comment]
@jbednar @philippjfr Happy with this workaround fix for now?

File: param/__init__.py [hunk 2]
```diff
@@ -1359,10 +1359,11 @@ class List(Parameter):
...
     """
 
-    __slots__ = ['class_','bounds']
+    __slots__ = ['bounds', 'item_type', 'class_']
 
-    def __init__(self,default=[],class_=None,instantiate=True,
+    def __init__(self,default=[],class_=None,item_type=None,instantiate=True,
                  bounds=(0,None),**params):
+        self.item_type = item_type or class_
         self.class_ = class_
```
Comment (philippjfr) [inline param/__init__.py:1367]: For backward compatibility this should also reflect the `item_type`. Is it possible to simply make `class_` a property?

Comment (philippjfr) [reply param/__init__.py:1367]: Just as an example, this change will already break the serialization code if a user uses item_type rather than `class_`.

Comment (jlstevens) [reply param/__init__.py:1367]: > Is it possible to simply make `class_` a property?

I don't think so...I hit an issue when importing holoviews where `class_` as a slot was expected which is why I re-added it. What I should do is keep them in sync, so I'll push this change:

```python
self.class_ = item_type or class_
```

Which should fix the serialization issue.

Comment (jlstevens) [reply param/__init__.py]: Addressed in 9096e39

File: param/__init__.py [hunk 3]
```diff
@@ -1359,11 +1359,12 @@ class List(Parameter):
...
 
-    __slots__ = ['class_','bounds']
+    __slots__ = ['bounds', 'item_type', 'class_']
 
-    def __init__(self,default=[],class_=None,instantiate=True,
+    def __init__(self,default=[],class_=None,item_type=None,instantiate=True,
                  bounds=(0,None),**params):
-        self.class_ = class_
+        self.item_type = item_type or class_
+        self.class_ = item_type or class_
```
Comment (philippjfr) [inline param/__init__.py:1367]:
```suggestion
        self.class_ = self.item_type
```

File: param/__init__.py (hunk 2 above)
Comment (jbednar) [reply param/__init__.py]: Won't there still be an issue when deserializing between different param versions, such that class_ was serialized previously but gets deserialized only to class_ and not iterm_type?  Seems like there needs to be the same sort of logic in get_state.

Comment (jlstevens) [reply param/__init__.py]: Annoying but yes, good point!

**Comment by jlstevens** [comment]
@philippjfr I see you've merged but what about Jim's point about pickling and `__getstate__` needing updating?

**Comment by philippjfr** [comment]
Missed that :/

**Pull request merged**